### PR TITLE
implement: json to sublime.decode_value/encode_value conversion

### LINKED ISSUE #119: "No JSON object could be decoded" when downloading gists
Hi, 

Both computers are running Ubuntu 16.04. 

I am attempting to download settings from Computer A on Computer B. 

On B, I moved my previous sublime-text profile to a separate folder so I could start from scratch. 

On B, I opened sublime, installed package manager and Sync Settings; and closed sublime text.

I reopened sublime, created a new gist token for B, copied it 
into B's syncSettings file; 
I also copied the gist_id into B's sync settings file. 

B's syncsettings file can be found at 
https://gist.github.com/skorasaurus/3226b2c0458a4b0fa28792f6b1ebbe97#file-b-syncsettings-sublime

It is unclear whether I need to include an access_token on each computer that you plan to sync settings with and whether that access token is the same for each computer. I understand that the gist_id should be each computer's syncsettings. 

I restarted sublime again and when I ran Sync Settings Download from Computer B, I received a modal dialog box pop-up:

"No JSON object could be decoded".

Lastly, I should note, in my original gist that I created on computer A, I receive a message at the top of the gist stating:
" This gist exceeds the recommended number of files (~10). To access all files, please clone this gist. To share more, create a repository. "

May this be the cause of the problem? 

I also included the package-control-sublime-user-settings and sync settings log from computer B.

### LINKED ISSUE #144: No JSON object could be decoded
Hey! Faced a problem.
Maybe it correlates with the new version of sublime text 4070 alpha

image

```
WARNING:Sync Settings.sync_settings.libs.logger:{'message': 'Not Found', 'documentation_url': 'https://developer.github.com/v3/gists/#delete-a-gist'}
ERROR:Sync Settings.sync_settings.libs.logger:No JSON object could be decoded
Traceback (most recent call last):
  File "./python3.3/json/decoder.py", line 367, in raw_decode
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\krupi\AppData\Roaming\Sublime Text 3\Installed Packages\Sync Settings.sublime-package\sync_settings/commands/download.py", line 53, in download
    package_settings = json.loads('{}' if file_content == '' else file_content)
  File "./python3.3/json/__init__.py", line 316, in loads
  File "./python3.3/json/decoder.py", line 351, in decode
  File "./python3.3/json/decoder.py", line 369, in raw_decode
ValueError: No JSON object could be decoded
ERROR:Sync Settings.sync_settings.libs.logger:No JSON object could be decoded
Traceback (most recent call last):
  File "./python3.3/json/decoder.py", line 367, in raw_decode
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\krupi\AppData\Roaming\Sublime Text 3\Installed Packages\Sync Settings.sublime-package\sync_settings/commands/download.py", line 53, in download
    package_settings = json.loads('{}' if file_content == '' else file_content)
  File "./python3.3/json/__init__.py", line 316, in loads
  File "./python3.3/json/decoder.py", line 351, in decode
  File "./python3.3/json/decoder.py", line 369, in raw_decode
ValueError: No JSON object could be decoded
```

My gist 
https://gist.github.com/krupitskas/b272fea836faffa356c8f0cce9a121b3

### PR DESCRIPTION
**Pull request by TheSecEng**
fixes #144
fixes #119

**BLUF**
Python's json library doesn't allow for trailing comments, or comments 🤷‍♂️ because it follows JSON Spec. Sublime files however often time have these, as sublime allows them. So too accommodate for that, we use sublimes builtin encode_value and decode_value for handling json.

@mfuentesg Not sure how to implement test case. Sorry, I've not learned python 100%. Could use some general guidance.

### CODE REVIEW

**Comment by mfuentesg** [comment]
@TheSecEng you need to create mock for `encode_value` and `decode_value` functions. Take a look to this [file](https://github.com/mfuentesg/SyncSettings/blob/master/tests/mocks/sublime_mock.py).

One way to test this change is, is implementing json.loads in `sublime.decode_value`.

For example here,  i am mocking the behavior of `sublime.yes_no_cancel_dialog` function, try to do the same with `sublime.decode_value`.

https://github.com/mfuentesg/SyncSettings/blob/a75694803ecee02a463f545ef3754b537f8ed70c/tests/test_sync_version.py#L45-L52

**Comment by mfuentesg** [comment]
@TheSecEng , please try to run your tests locally, before to push. Thanks in advance for your contribution.

**Comment by TheSecEng** [comment]
@mfuentesg thanks for some direction, looks like I have that portion working, not sure how to modify the `get_local_version` tests to pass since they are using sublime functions

**Comment by mfuentesg** [comment]
@TheSecEng i will check it after my work hours.

**Comment by mfuentesg** [comment]
i'll try to help you to write tests

**Comment by TheSecEng** [comment]
I want to add that the following functions were adding in Sublime Text 3019, likely not a problem.

`sublime.encode_value`
`sublime.decode_value`

**Comment by TheSecEng** [comment]
@mfuentesg I am sorry I can't help more with writing the tests. Have you been able to look at this more?

**Comment by mfuentesg** [comment]
Hey @TheSecEng thanks for your time, i will try to take it next week.

**Comment by dalisoft** [comment]
Who doesn't wait or need currently this feature, do these feature

_This feature works only in Unix systems (macOS, Linux, WSL)_

1. Remove current Sync Settings (Disabling doesn't work)
2. Press **Cmd**/**Ctrl + Shift + P**, search **Browser Packages** and click Enter
3. Copy path and paste go-to directory via `cd DIR` or Open terminal here (available for Linux)
4. Run `git clone https://github.com/TheSecEng/SyncSettings.git && cd SyncSettings` in Terminal
5. Run `git checkout fix/json-handling`
6. Restart Sublime Text
7. Press **Cmd**/**Ctrl + Shift + P** again, but now search **Satisfy Dependencies** and click Enter
8. Restart Sublime Text

Done, you should get working solution

Thanks to
@mfuentesg - for original best package, thanks billion times
@TheSecEng - for fork and fix, thanks billion times
and to all contributors of SyncSettings package + Sublime Text teams for fast editor

**Comment by TheSecEng** [comment]
> Who doesn't wait or need currently this feature, do these steps
> 
> _This feature works only in Unix systems (macOS, Linux, WSL)_
> 
> 1. Remove current Sync Settings (Disabling doesn't work)
> 2. Press **Cmd**/**Ctrl + Shift + P**, search **Browser Packages** and click Enter
> 3. Copy directory path and paste go-to directory via `cd DIR` or Open terminal here at directory via context menu (available for Linux)
> 4. Run `git clone https://github.com/TheSecEng/SyncSettings.git` in Terminal
> 5. Run `cd SyncSettings && git checkout fix/json-handling` in Terminal
> 6. Restart Sublime Text
> 7. Press **Cmd**/**Ctrl + Shift + P** again, but now search **Satisfy Dependencies** and click Enter
> 8. Restart Sublime Text
> 
> Done, you should get working solution
> 
> Thanks to
> @mfuentesg - for original best package, thanks billion times
> @TheSecEng - for fork and fix, thanks billion times
> and to all contributors of SyncSettings package + Sublime Text teams for fast editor
> 
> Regards,
> Davlik

I highly suggest not doing this. I will not be maintaining my branch and it will be deleted once this is merged.

**Comment by dalisoft** [comment]
@TheSecEng after this is merged, we will have original package with fixes and these steps does not need. But thanks for warning

**Comment by mfuentesg** [comment]
@dalisoft @TheSecEng thank you guys, i am very busy with my job, i haven't the bandwidth yet.
Give me some time please :D.

Sorry for the inconvenience

**Comment by mfuentesg** [comment]
@TheSecEng plase find my comments to solve the issue.

File: tests/mocks/sublime_mock.py [hunk 1]
```diff
@@ -29,3 +29,11 @@ def load_settings(*args):
         'included_files': [],
         'excluded_files': []
     })
+
+
+def encode_value(*args):
+    pass
+
+
+def decode_value(*args):
```
Comment (mfuentesg) [inline tests/mocks/sublime_mock.py:38]: replace this function by:

```python
def decode_value(content):
    decoded = re.sub(re.compile("/\*.*?\*/",re.DOTALL ) ,"" ,content)
    return json.loads(re.sub(re.compile("//.*?\n" ) ,"" ,decoded))
```


With this you can replicate the behavior of "sublime.decode_value"

File: tests/mocks/sublime_mock.py [hunk 2]
```diff
@@ -29,3 +29,11 @@ def load_settings(*args):
         'included_files': [],
         'excluded_files': []
     })
+
+
+def encode_value(*args):
```
Comment (mfuentesg) [inline tests/mocks/sublime_mock.py:34]: replace this function by:

```python
def encode_value(data):
    return json.dumps(data)
```

With this, you will be able to replicate the behavior of `sublime.encode_value`

File: tests/mocks/sublime_mock.py (hunk 1 above)
Comment (mfuentesg) [reply tests/mocks/sublime_mock.py:38]: this function removes comments from any json content.

Comment (mfuentesg) [reply tests/mocks/sublime_mock.py:38]: Can you add a new test validating this behavior?

something like this.

```python
    @mock.patch('sync_settings.libs.path.exists', mock.MagicMock(return_value=True))
    @mock.patch('sync_settings.sync_version.open', mock.mock_open(
        read_data='{"created_at": "2019-01-11T02:15:15Z", /* some comment */"hash": "123123123"}'))
    def test_get_local_version_with_commented_content(self):
        v = version.get_local_version()
        self.assertDictEqual({'hash': '123123123', 'created_at': '2019-01-11T02:15:15Z'}, v)
```

**Comment by mfuentesg** [comment]
@TheSecEng apart of the requested changes, please add yourself to the contributor list in readme file.

**Comment by mfuentesg** [comment]
@allcontributors please add @TheSecEng

**Comment by mfuentesg** [comment]
@allcontributors please add @TheSecEng for code

**Comment by mfuentesg** [comment]
@allcontributors please add @TheSecEng  for code

**Comment by TheSecEng** [comment]
@mfuentesg Thanks, all is good finally.

**Comment by mfuentesg** [comment]
@allcontributors please add @TheSecEng for code

**Comment by dalisoft** [comment]
As #167 is merged, seems, this [way](https://github.com/mfuentesg/SyncSettings/pull/151#issuecomment-647148634) is only one to get it working on ST4 on macOS?

**Comment by mfuentesg** [comment]
@dalisoft those changes are already deployed and tagged, using latest version of ST should has this fixed.

**Comment by dalisoft** [comment]
@mfuentesg  I know this changed already merged, but also reverted via #167, isn't? I love ST and amazing plugins made by community, but it's not working, only way to get it working is same comment as linked above.
Thanks a lot for your hard work to make this plugin

**Comment by dalisoft** [comment]
image

My gist settings is [here](https://gist.github.com/dalisoft/53d864ba0cd20ccd405e8aa4d741ab85#gistcomment-3560779)

**Comment by mfuentesg** [comment]
@dalisoft please find issue here. https://github.com/mfuentesg/SyncSettings/issues/185

**Comment by dalisoft** [comment]
@mfuentesg Thanks

**Comment by mfuentesg** [comment]
@dalisoft https://github.com/mfuentesg/SyncSettings/issues/185 this is already merged, please let me know if it solve your issue 👍🏻

**Pull request merged**